fxa-oauth-server/lib/routes/token.js

```diff
@@ -78,17 +75,23 @@ const PAYLOAD_SCHEMA = Joi.object({
       is: GRANT_REFRESH_TOKEN,
       then: Joi.optional()
     })
+    .when('grant_type', {
+      is: GRANT_FXA_ASSERTION,
```

Comment:
1. a buffer is returned for incorrect secret, like so:

![](http://v14d.com/i/447ffe767a4414262f26eea279271e81d0c6706e03e48f8ff72a3b1a6a710ab4.jpg)

2.  with a client that is not a public client , the error produces info that `client_secret` is invalid. Should it reject a non-public client? 

![](http://v14d.com/i/e4fdac7b6d874a42a7bdf001c5f495549fbd1b2ad928e2a88986e051c783bc72.jpg)

Reply:
> a buffer is returned for incorrect secret, like so:

This appears to be a side-effect of the previous refactor of this endpoint, which removed is heavy use of automagical bytes-to-hex conversion.  I've pushed a second commit to this PR which casts byte identifiers to hex when constructing error objects, to avoid this problem in future.

> with a client that is not a public client , the error produces info that client_secret is invalid.

This behavior seems accurate to me, you're attempting to act as a non-public client without providing the corresponding client secret, so `authenticateClient` is rejecting your request before it even gets to deciding whether you're allowed to use the `fxa-assertion` grant type.

If you provided the correct `client_secret`, I expect this would instead return an "invalid grant type" error, because that client id does not have `canGrant` permission.

> Should it reject a non-public client?

It will reject any clients with `canGrant: false`. I can't think of any reason why we'd ever have a non-public client with `canGrant: true`, so perhaps it's worth adding an extra check for this as a safety measure.  I'll update.

(But as above, I think it's fine to only do this check *after* you've provided the correct client_secret for that client_id).